signac/core/collection_api.py

```diff
@@ -0,0 +1,446 @@
+
+    def __eq__(self, other):
+        if isinstance(other, type(self)):
+            return self() == other()
+        else:
+            return self() == other
+
+    def __len__(self):
+        self.load()
+        return len(self._data)
+
+    def keys(self):
+        self.load()
+        return self._data.keys()
+
+    def values(self):
+        self.load()
+        return self.to_base().values()
+
+    def items(self):
+        self.load()
+        return self.to_base().items()
+
+    def get(self, key, default=None):
+        self.load()
+        return self._data.get(key, default)
+
+    def pop(self, key, default=None):
+        self.load()
+        ret = self._data.pop(key, default)
+        self.sync()
+        return ret
+
+    def popitem(self, key, default=None):
+        self.load()
+        ret = self._data.pop(key, default)
+        self.sync()
+        return ret
+
+    def clear(self):
+        self._data = {}
+        self.sync()
+
+    def update(self, mapping):
+        self.load()
+        with self._suspend_sync():
+            for key, value in mapping.items():
+                self[key] = self.from_base(value, self)
+        self.sync()
+
+    def setdefault(self, key, default=None):
+        self.load()
+        with self._suspend_sync():
+            ret = self._data.setdefault(key, self.from_base(default, self))
+        self.sync()
+        return ret
+
+
+class SyncedDict(_SyncedDict):
+
+    _PROTECTED_KEYS = ('_data', '_suspend_sync_', '_load', '_sync', '_parent')
+
+    def __getattr__(self, name):
+        try:
+            return super(SyncedDict, self).__getattribute__(name)
+        except AttributeError:
+            if name.startswith('__'):
+                raise
+            try:
+                return self.__getitem__(name)
+            except KeyError as e:
+                raise AttributeError(e)
+
+    def __setattr__(self, key, value):
+        try:
+            super(SyncedDict, self).__getattribute__('_data')
+        except AttributeError:
+            super(SyncedDict, self).__setattr__(key, value)
+        else:
+            if key.startswith('__') or key in self.__getattribute__('_PROTECTED_KEYS'):
+                super(SyncedDict, self).__setattr__(key, value)
+            else:
+                self.__setitem__(key, value)
+
+    def __delattr__(self, key):
+        if key.startswith('__') or key in self._PROTECTED_KEYS:
+            super(SyncedDict, self).__delattr__(key)
+        else:
+            self.__delitem__(key)
+
+
+class SyncedList(SyncedCollection, MutableSequence):
+
+    def __init__(self, data=None, parent=None):
+        super(SyncedList, self).__init__(parent=parent)
+        if data is None:
+            self._data = []
+        else:
+            self._data = [self.from_base(value, self) for value in data]
+
+    def to_base(self):
+        """Recursively crawl a SyncedList and convert all elements to
+        corresponding synced elements."""
```

Comment:
This docstring sounds a little misleading to me. The conversion is to _base_ (unsynced) collections, not synced elements.

Reply:
Okay, will change it.

Reply:
@vishav1771 it looks like you just removed the docstring, are you planning to replace this?

Reply:
Yeah, I was planning to rewrite so I removed for now.